### PR TITLE
Add dataframe datatype as alternative to csv and tabular

### PR DESCRIPTION
**Pull request by qiagu**
## What did you do? 
- [describe the proposed changes]
- Add dataframe datatype

## Why did you make this change?
(Cite Issue number OR provide rationalization of changes if no issue exists)
(If fixing a bug, please add any relevant error or traceback)

## How to test the changes? 
(select the most appropriate option; if the latter, provide steps for testing below)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

## For UI Components
- [ ] I've included a screenshot of the changes

### CODE REVIEW

**Comment by qiagu** [comment]
image

**Comment by mvdbeek** [comment]
Can you modify the tabular dataset peek please ? I'd suggest just skipping the peek beyond a certain dataset size. I don't think this should be a separate datatype. Longer term we want to completely generate the peek client side, I think.

**Comment by qiagu** [comment]
@mvdbeek  Thanks for your comments. What you said makes sense, but please let me finish most of my idea before judging whether it worths a separate datatype or not. Also, I think being able to peek small part of the dataset, even as small as 5 x 5 , is still valuable.

**Comment by qiagu** [comment]
The code is still rough, but main idea has been implemented. As a backup/alternative datatype, it doesn't need a sniff method.
@mvdbeek  what do you think now?

**Comment by mvdbeek** [comment]
I don't think this is necessary. You can limit the number of columns in the peek (we really have to do that!) in the tabular datatype. Having a separate datatype provides no advantage that I can see.

**Comment by dannon** [comment]
@mvdbeek We won't want to generate the peek client-side for many datatypes, if any -- that'd mean file IO on peek unless the peek is purely metadata-based.  As an intentional specialization of tabular with potentially different tools available, validation, metadata, display methods, etc. I don't think it's unreasonable.

**Comment by mvdbeek** [comment]
> that'd mean file IO on peek unless the peek is purely metadata-based

Yes, that is what it should be IMO. I don't think shipping html like this is a good idea, but that's a different discussion. None of what I'm seeing here couldn't be implemented in our standard tabular datatype. If we add another datatype here we're causing a lot of pain to our users and tool developers for no good reason.

**Comment by dannon** [comment]
I agree that shipping raw html hasn't been a good idea for a while now, but there are real benefits to being able to ship a data preview without slowing the transaction down with file IO.  Right now we have html, but I would imagine in the future we'll just want to ship a json blob `raw_peek`, or the like, that the client handles appropriately without touching the filesystem.  Anyway, you're right that's a completely separate thing.

On the datatype itself -- I feel like we specialize datatypes all the time (even if it's just a named subclass) and this has been a particular pain point for folks on the ML/etc side for years.  Is this one much worse for reasons I'm not seeing?

**Comment by mvdbeek** [comment]
I mean, sure, when we specialize a datatype I fully agree and I've added my share of datatypes, but this is a general purpose datatype that does exactly the same as the tabular datatype, it fixes a bug (https://github.com/galaxyproject/galaxy/issues/11885) in a roundabout way that we should fix in the tabular datatype.

`tabular` is without a doubt the most used datatype in Galaxy, I would be really sad to navigate yet another datatype that has the exact same specifications that tabular has, and that works in every tool that accepts tabular. From the user perspective I can't imagine a reason to do this specialization (that is not a specialization). There's no tool that can accept tabular but not dataframe, or vice versa.

 There's the other minor point that pandas is not a lightweight dependency, but okay, that's not the main thing I'm concerned with.

**Comment by qiagu** [comment]
This `dataframe`  indeed is a data science-flavored datatype.  As an extra benefit, it simplifies choice b/w `csv` and `tabular`.  The `tabular` collects rich metadata, which seems unnecessary for routine pandas operators. Also, as it is so widely used in Galaxy, I am not sure I have enough knowledge to make it right. These are the factors that push me to make a new datatype. I like to see some tweak on `tabular` as well.

File: lib/galaxy/datatypes/tabular.py [hunk 1]
```diff
@@ -1380,3 +1381,75 @@ def set_meta(self, dataset, overwrite=True, skip=None, max_data_lines=7, **kwd):
...
+        if offset is not None:
+            return self.get_chunk(trans, dataset, offset, ck_size)
+        elif to_ext or not preview:
+            to_ext = to_ext or dataset.extension
+            return self._serve_raw(trans, dataset, to_ext, **kwd)
+
+        # suppose no out of index range
+        start_column = columns_per_page * page
+        end_column = min(start_column + columns_per_page, dataset.metadata.columns)
+        df = pd.read_csv(
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/tabular.py]: This reads an unbounded amount of data into memory. You're limiting the rows and columns, but not the size of individual fields. You may get around this by reading a fixed-size fragment into a StringIO instance and calling `pd.read_csv` on that, but that will break when you're reading the next chunk. 

This branch also doesn't seem right, this code branch should return an initial chunked view so you can request more chunks. Now there's no indication that there is actually more data in the file (which would be fine if you don't want to implement chunks), it just isn't shown.

File: lib/galaxy/datatypes/tabular.py [hunk 2]
```diff
@@ -1380,3 +1381,75 @@ def set_meta(self, dataset, overwrite=True, skip=None, max_data_lines=7, **kwd):
...
+    def get_shape(self, dataset):
+        return (dataset.metadata.data_lines, dataset.metadata.columns)
+
+    def set_peek(self, dataset, peek_max_lines=5, peek_max_columns=10):
+        if not dataset.dataset.purged:
+            df = pd.read_csv(
+                dataset.file_name, sep=dataset.metadata.delimiter,
+                usecols=range(min(peek_max_columns, dataset.metadata.columns)),
+                nrows=peek_max_lines)
+            dataset.peek = df.to_json()
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/tabular.py:1457]: This has the same problem that the `display_data` method below has, the json here can be arbitrarily large and take up a large amount of space on the peek column of the history_dataset_association table. The StringIO workaround here might be a good one though, and i like the idea of storing the peek in JSON, although I am not sure that JSON is the most efficient way to store tabular data, maybe csv would be better here.

File: lib/galaxy/datatypes/tabular.py (hunk 1 above)
Comment (qiagu) [reply lib/galaxy/datatypes/tabular.py]: @mvdbeek The unbounded data is a really good point. I don't have good solution for now. 
For the chunks, yes, there are no chunks in current implementation, probably won't make ones from the columns . However, chunks can be made from iteration of the rows.

File: lib/galaxy/datatypes/tabular.py (hunk 2 above)
Comment (qiagu) [reply lib/galaxy/datatypes/tabular.py:1457]: I can try the `to_string` method here.

File: lib/galaxy/datatypes/tabular.py (hunk 1 above)
Comment (qiagu) [reply lib/galaxy/datatypes/tabular.py]: @mvdbeek I wonder what's the real concern about the data size limit. If we care more about the amount of data transferred to client but not the peak memory usage in computation, the size of dataframe object can be shrunk by using parameter `converters` (e.g., long string can be truncated). If memory usage in computation / IO is a big concern, we probably have to closely monitor each `readline()` call in the pandas parser. In the latter case, I think current `tabular` implementation is not completely safe.

Comment (qiagu) [reply lib/galaxy/datatypes/tabular.py]: IMHO, fixed number of cells plus `converters` works for most regular use cases.

File: lib/galaxy/datatypes/tabular.py (hunk 2 above)
Comment (qiagu) [reply lib/galaxy/datatypes/tabular.py:1457]: The `to_string` seems worse than `to_json`. :(

File: lib/galaxy/datatypes/tabular.py (hunk 1 above)
Comment (qiagu) [reply lib/galaxy/datatypes/tabular.py]: @mvdbeek  The StringIO method rocks!

Comment (qiagu) [reply lib/galaxy/datatypes/tabular.py]: My plan was to make page navigator on UI. One page in current setting is 50 KB at most. No chunk is needed?

**Comment by qiagu** [comment]
Which is better, continuous scrolling vs. manual paginated?
image

File: lib/galaxy/datatypes/tabular.py (hunk 2 above)
Comment (mvdbeek) [reply lib/galaxy/datatypes/tabular.py:1457]: That's why I suggested csv (using `to_csv()`)

**Pull request closed without merging**